test/e2e/apps/statefulset.go

```diff
@@ -1400,6 +1464,73 @@ var _ = SIGDescribe("StatefulSet", func() {
 			err = verifyStatefulSetPVCsExist(ctx, c, ss, []int{0})
 			framework.ExpectNoError(err)
 		})
+
+		ginkgo.It("should not delete PVCs when there is another controller", func(ctx context.Context) {
+			e2epv.SkipIfNoDefaultStorageClass(ctx, c)
+			ginkgo.By("Creating statefulset " + ssName + " with no retention policy in namespace " + ns)
+
+			*(ss.Spec.Replicas) = 3
+			_, err := c.AppsV1().StatefulSets(ns).Create(ctx, ss, metav1.CreateOptions{})
+			framework.ExpectNoError(err)
+
+			ginkgo.By("Confirming all 3 PVCs exist")
+			err = verifyStatefulSetPVCsExist(ctx, c, ss, []int{0, 1, 2})
+			framework.ExpectNoError(err)
+
+			claimNames := make([]string, 3)
+			for i := 0; i < 3; i++ {
+				claimNames[i] = fmt.Sprintf("%s-%s-%d", statefulPodMounts[0].Name, ssName, i)
+			}
+
+			ginkgo.By("Add external controller to PVC 1")
+			claim, err := c.CoreV1().PersistentVolumeClaims(ns).Get(ctx, claimNames[1], metav1.GetOptions{})
+			framework.ExpectNoError(err)
+			claim.SetOwnerReferences(append(claim.GetOwnerReferences(), metav1.OwnerReference{
+				APIVersion: "unknown/v1",
+				Kind:       "controller",
+				Name:       "unknown",
+				UID:        "unknown",
+			}))
+			_, err = c.CoreV1().PersistentVolumeClaims(ns).Update(ctx, claim, metav1.UpdateOptions{})
+			framework.ExpectNoError(err)
+
+			ss, err = c.AppsV1().StatefulSets(ns).Get(ctx, ssName, metav1.GetOptions{})
+			framework.ExpectNoError(err)
+			ginkgo.By("Update retention policy to delete")
+			ss.Spec.PersistentVolumeClaimRetentionPolicy = &appsv1.StatefulSetPersistentVolumeClaimRetentionPolicy{
+				WhenDeleted: appsv1.DeletePersistentVolumeClaimRetentionPolicyType,
+			}
+			ss, err = c.AppsV1().StatefulSets(ns).Update(ctx, ss, metav1.UpdateOptions{})
+			framework.ExpectNoError(err)
+
+			ginkgo.By("Expect claims 0 and 2 to have ownerRefs to the statefulset")
+			err = wait.PollUntilContextTimeout(ctx, e2estatefulset.StatefulSetPoll, e2estatefulset.StatefulSetTimeout, true, func(ctx context.Context) (bool, error) {
+				for i := range []int{0, 2} {
+					claim, err := c.CoreV1().PersistentVolumeClaims(ns).Get(ctx, claimNames[i], metav1.GetOptions{})
+					if err != nil {
+						return false, nil // retry
+					}
+					found := false
+					for _, ref := range claim.GetOwnerReferences() {
+						if ref.Kind == "StatefulSet" && ref.Name == ssName {
+							found = true
+							break
+						}
+					}
+					if !found {
+						return false, nil // retry
+					}
+				}
+				return true, nil // found them both!
+			})
+			framework.ExpectNoError(err)
```

Comment:
can we remove the controller flag from one of the claims here and check it gets reconciled back to true?

Reply:
Done